### PR TITLE
fix(protocol): fix 2 bridge bugs reported by Victor

### PR DESCRIPTION
**Pull request by dantaik**
Reported by Victor from QuillAudits:

1.  if message.to == etherVault bug, then message can call releaseEther to drain all Ether from the audit.

2. currently attacker can steal all tokens in the ERC20Vault through the same bug I mentioned, calling the onMessageRecalled which also has a constraint that shouldn't be called unless failed, currently through this bug that is not handled fully attacker is able to bypass this regardless and steal all tokens

### CODE REVIEW

File: packages/protocol/contracts/bridge/Bridge.sol
Comment (Brechtpd) [inline packages/protocol/contracts/bridge/Bridge.sol:212]: Is this only added here for messages that were already sent before adding the new checks? Just want to make sure I don't miss anything.

Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:212]: Without this, we'd have to rely on all source AddressManagers to be configured correctly; with this, the Bridge is considered more secure.

**Review by Brechtpd — APPROVED** [review]

File: packages/protocol/contracts/bridge/Bridge.sol
Comment (dantaik) [inline packages/protocol/contracts/bridge/Bridge.sol:101]: @Brechtpd Here, maybe I can simply remove this if-statement.

Comment (Brechtpd) [reply packages/protocol/contracts/bridge/Bridge.sol]: Ah yeah, that makes sense!

**Review by Brechtpd — APPROVED** [review]

**Review by davidtaikocha — APPROVED** [review]
LGTM, after merging this into `4844`, lets open another cherry-pick PR targeting to `contestable-zkrollup`?

File: packages/protocol/contracts/bridge/Bridge.sol
Comment (Brechtpd) [inline packages/protocol/contracts/bridge/Bridge.sol:156]: I guess for `onMessageRecalled` the context could simply be passed into the function because of the fixed function call, but makes sense to use the same mechanism as the normal call.

**Review by Brechtpd — APPROVED** [review]

File: packages/protocol/contracts/bridge/Bridge.sol
Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Not sure I understand you correctly. I think vaults must call back to the Bridge through `context()` to make sure the onMessageCalled is not only called by bridge, but only called within the `recallMessage` function.

Comment (Brechtpd) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Isn't that special case handled by having the `assert(message.from != address(this));` in the normal `_invokeMessageCall`? So the context from the bridge can only contain the bridge address itself for `recallMessage`?

Not suggesting any changes, but still interesting to know to make sure I fully understand things. :)

Comment (Brechtpd) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Oh never mind, I see know why that can't work!

Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: In `sendMessage`, we have `_message.from = msg.sender;` this means message.from should never be "address(this)" -- so the assert will never be triggered.

 But a user can still specify `message.to` to a vault, and the `message.data= abi.encode(Vault.onMessageRecalled.selector, ...)`.  Since the vault trusts the Bridge -- its `recallMessage` function has a `onlyFromNamed("bridge")` modifier, the `onMessageRecalled` can be executed if the user process the message on the destination.

So any bridge-enabled app, all the current token vaults included, must verify that certain functions are not only callable by the bridge, but only by the bridge's processMessage or recallMessage function, not both.

Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: The security is so important, do you think we should set up a call to discuss about it?

**Comment by dantaik** [comment]
> LGTM, after merging this into `4844`, lets open another cherry-pick PR targeting to `contestable-zkrollup`?

David, there are actually some outstanding issues with multi-hop bridging, and its not really easy to resolve soon.

File: packages/protocol/contracts/bridge/Bridge.sol
Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Also, I realized some design assmptions are not valid for enabling the sharing of certain components, such as EtherVault, token vaults, signal service, the bridge, among multiple L2s. To truly make L2->L2 transfer of asset possible, there may be quite some changes... Still working on it.

Comment (Brechtpd) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: > The security is so important, do you think we should set up a call to discuss about it?

Looking at these issues, I realized I'm really not familiar with how the bridge works, so for me I think it would be best to first spend some more time learning the latest code.

If there's more changes coming, maybe we can do a call after everybody had time to look into those? Or we need a bridge session in Turkey. :)

Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Ok, there will be more changes, mostly to:

1) simplify the vault codebase
2) enable L2-L2 bridging (still not working)

Comment (dantaik) [reply packages/protocol/contracts/bridge/Bridge.sol:156]: Let me merge this one first.

**Pull request merged**
